### PR TITLE
add Code component for showing/copying snippets

### LINKED ISSUE #244: Add CodeBlock component
Used in a bunch of places in Weave Cloud.

Should support both shell commands and raw file format, with standardized coloring.

image

### PR DESCRIPTION
**Pull request by guyfedwards**
I have tried to simplify the API and usage to be more intuitive although for more complex styled/dynamic setups there is still some jsx wrangling required.

Couple of things to discuss:
 - current implementation has the content scroll when it overflows. I think this is the sanest option, but visually will change across different OS/Browsers?
 - Do we want to shrink the copy button so it occupies less of the code space?

image

image

image

fixes #244

### CODE REVIEW

**Comment by fbarl** [comment]
> Current implementation has the content scroll when it overflows. I think this is the sanest option, but visually will change across different OS/Browsers?

I think this is good enough for now.

> Do we want to shrink the copy button so it occupies less of the code space?

I just talked about it with @bia and we both think it might be better if we just got rid of that button :) What we could have instead is some sort of a popup button on text selection - I'll let her post some screenshots.

That way we would solve both the vertical spacing and potentially partial selection issue (see https://github.com/weaveworks/service-ui/issues/2696).

Nice examples btw! Will take a look at the code once we settle on the _Copy_ button.

**Comment by guyfedwards** [comment]
👍 Cool beans. Agree about the button, I had been wondering about something like this:
normal:
image
hover:
image

**Comment by bia** [comment]
Sounds good.
I was talking to @fbarl about even showing no overlay at all, then on mouse down the word 'Copied' or 'Copied to clipboard' in the top right corner. Then the person could continue to select normally within the code block if they needed one piece in particular.

What do you think? Shall I come up with a mockup tomorrow?

**Comment by bia** [comment]
How does this feel?

copycode
copycode2

I would also keep the slightly rounded corners - it helps a block this dark not to seem so visually heavy.

**Comment by guyfedwards** [comment]
@bia Rounded corners 👍 forgot to add them back, couple of questions: 
- Is that triggered on clicking the code block itself? (are we implementing the partial selection as per https://github.com/weaveworks/service-ui/issues/2696)
- Is it obvious to the user that thats how it works 🤔 ?

**Comment by bia** [comment]
>Is that triggered on clicking the code block itself? (are we implementing the partial selection as per weaveworks/service-ui#2696)

We are going to bring back partial selection.

>Is it obvious to the user that thats how it works 🤔, some indication that they can click it, an icon maybe?

Good point.
What if we display this as a hover state
copycode3

and this on mouse down.
copycode2


You could continue to drag and partially select to manually copy a portion, but at this point all of the text is in your clipboard.

**Comment by guyfedwards** [comment]
You can now click anywhere on the code block to copy all the text, or select some text and only that text will be selected.

File: src/components/Code/example.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import React from 'react';
+import { trim } from 'lodash';
+import styled from 'styled-components';
+
+import { Example, Info } from '../../utils/example';
+import { Input } from '..';
+import Code from '.';
+
+const Highlight = styled.span`
+  color: ${props => props.theme.colors.accent.blue};
```
Comment (fbarl) [inline src/components/Code/example.js]: Please rebase this PR to capture the latest color naming changes.. Sorry for that :/

File: src/components/Code/example.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            kubectl create secret generic -n weave cloudwatch
+            --from-literal=access-key-id=xyz
+            --from-literal=secret-access-key=abc
+          </Code>
+        </Example>
+
+        <Info>Multiline file</Info>
+        <Example>
+          <Code isFile onCopy={this.onCopy}>
+            {trim(`
```
Comment (fbarl) [inline src/components/Code/example.js]: Perhaps it would make sense to wrap `children` in `trim` internally if the string is passed as the only param, i.e. `isString(children)`.

File: src/components/Code/example.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  --from-literal=access-key-id=xyz \\
+  --from-literal=secret-access-key=abc
+          `)}
+          </Code>
+        </Example>
+
+        <Info>Dynamic content in template string</Info>
+        <Example>
+          <Code isFile onCopy={this.onCopy}>
+            {() =>
```
Comment (fbarl) [inline src/components/Code/example.js]: Is there a use case for passing a function? In a sense that if it wasn't a function, something would not work (or be much more complicated)?

File: src/components/Code/example.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                {'}'}
+              </div>
+              <div>{'}'}</div>
+            </div>
+          </Code>
+        </Example>
+      </div>
+    );
+  }
+}
```
Comment (fbarl) [inline src/components/Code/example.js]: Great examples! :+1:

File: src/components/Code/Code.test.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import Adapter from 'enzyme-adapter-react-16';
+import { withTheme } from '../../utils/theme';
+import Code from '.';
+
+enzyme.configure({ adapter: new Adapter() });
+
+document.execCommand = () => {};
+document.getSelection = () => ({
+  toString: () => '',
+});
```
Comment (fbarl) [inline src/components/Code/Code.test.js]: Do these get properly rolled back after the tests are finished?

File: src/components/Code/Code.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+const CopyNotice = styled.div`
+  position: absolute;
+  top: 5px;
+  right: 5px;
+  transition: opacity 300ms ease;
+  opacity: 0;
+  font-size: ${props => props.theme.fontSizes.tiny};
+  color: ${props => props.theme.colors.athensGray};
+`;
```
Comment (fbarl) [inline src/components/Code/Code.js]: I suggest a couple of styling changes here that IMO make it look better:

* Set both `top` and `right` to `0`
* Add padding `10px 15px` instead, like in the wrapper element
* Add the same `borderRadius` and `backgroundColor` like the wrapper element as well
* Set the full opacity to `0.9` instead of `1` below, so that the text overlay over the code looks a bit better
* Set font size to `small` instead of `tiny` after all

Basically, I think the text shouldn't be that much stuck to the corner and increasing the padding will potentially cause overlap with the text (which should be fine), therefore the rest of the changes.

cc @bia

File: src/components/Code/Code.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+const Pre = styled.pre`
+  margin: 10px 0;
+  font-family: ${props => props.theme.fontFamilies.monospace};
+  font-size: ${props => props.theme.fontSizes.small};
+`;
+
+class Code extends Component {
+  static displayName = 'Code';
```
Comment (fbarl) [inline src/components/Code/Code.js]: Wouldn't that already be a default?

File: src/components/Code/Code.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    } catch (e) {
+      throw e;
+    }
+    document.body.removeChild(txtArea);
+
+    // show the Copied to clipboard notice temporarily
+    this.setState({ isCopying: true });
+    setTimeout(() => {
+      this.setState({ isCopying: false });
+    }, 1000);
```
Comment (fbarl) [inline src/components/Code/Code.js]: I think I'd prefer this interval to be a bit longer - perhaps 3 seconds?

File: src/components/Code/Code.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+        <CopyNotice>
+          <i className={`fa fa-${copyIcon}`} /> {copyText}
+        </CopyNotice>
+      </Root>
+    );
+  }
+}
+
+export default Code;
```
Comment (fbarl) [inline src/components/Code/Code.js]: `propTypes` (and `defaultProps`) are missing here - we've decided to always add those in this repo, please see any other component around for how to do it.

File: src/components/Code/Code.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  color: ${props => props.theme.colors.athensGray};
+`;
+
+const Pre = styled.pre`
+  margin: 10px 0;
+  font-family: ${props => props.theme.fontFamilies.monospace};
+  font-size: ${props => props.theme.fontSizes.small};
+`;
+
+class Code extends Component {
```
Comment (fbarl) [inline src/components/Code/Code.js]: Component description (a comment above this line) that shows up in the _Examples_ page would also be nice.

File: src/components/Code/Code.test.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    );
+    expect(onCopy).not.toHaveBeenCalled();
+    wrapper.find('Code__CopyNotice').simulate('click');
+    expect(onCopy).toHaveBeenCalledTimes(1);
+  });
+
+  it('should render correctly', () => {
+    expect(wrapper.dive()).toMatchSnapshot();
+  });
+});
```
Comment (fbarl) [inline src/components/Code/Code.test.js]: Thanks for creating the tests! :ok_hand:

File: src/components/Code/Code.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  position: absolute;
+  top: 5px;
+  right: 5px;
+  transition: opacity 300ms ease;
+  opacity: 0;
+  font-size: ${props => props.theme.fontSizes.tiny};
+  color: ${props => props.theme.colors.athensGray};
+`;
+
+const Root = styled.div`
```
Comment (fbarl) [inline src/components/Code/Code.js]: `Root` is an odd name for an element (it gets shown in `className`) - normally we'd name this one `CodeWrapper` or something similar :)

File: src/components/Code/example.js (hunk 2 above)
Comment (guyfedwards) [reply src/components/Code/example.js]: Initially I had this but might there be an occasion where you want space before/after?

File: src/components/Code/example.js (hunk 3 above)
Comment (guyfedwards) [reply src/components/Code/example.js]: Allows for using template string formatting with dynamic values as opposed to having to deal with indentation in JSX

File: src/components/Code/Code.test.js (hunk 5 above)
Comment (guyfedwards) [reply src/components/Code/Code.test.js]: Great question, I believe they do as each test is isolated.
They are purely there to stop Jest barfing about `document.execCommand` not being a function as it is neither are mocked in JSDom.

Comment (guyfedwards) [reply src/components/Code/Code.test.js]: Just checked and the tests are isolated so this will not affect other tests regardless

File: src/components/Code/example.js (hunk 2 above)
Comment (fbarl) [reply src/components/Code/example.js]: I don't think so - if we currently don't have such a use case anywhere in the Service UI, I think you can assume it's not likely to happen :)

File: src/components/Code/example.js (hunk 3 above)
Comment (fbarl) [reply src/components/Code/example.js]: Ok I see, was wondering about the particular use case in Service UI, but found `FluxSetupKubectlCommand` :)

File: src/components/Code/Code.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+Code.propTypes = {
+  // Children can be anything that React can render
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
+
+  // onCopy will be called when the CodeWrapper is clicked
+  onCopy: PropTypes.func,
+};
+
+Code.defaultProps = {
+  onCopy: () => {},
```
Comment (fbarl) [inline src/components/Code/Code.js]: There is `lodash.noop` which is maybe a tiny bit cleaner :)

File: src/components/Code/Code.js [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+          <i className={`fa fa-${copyIcon}`} /> {copyText}
+        </CopyNotice>
+      </CodeWrapper>
+    );
+  }
+}
+
+Code.propTypes = {
+  // Children can be anything that React can render
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
```
Comment (fbarl) [inline src/components/Code/Code.js]: Hmm, so even strings are seen as `PropTypes.node`? Seems so, I don't see React complaining in the examples :)

Comment (guyfedwards) [reply src/components/Code/Code.js]: Yea node is anything that can be rendered, however functions aren't technically renderable until they are invoked so requires a separate type :man_shrugging:

File: src/components/Code/Code.js (hunk 13 above)
Comment (guyfedwards) [reply src/components/Code/Code.js]: certainly is :+1:
